### ISSUE TITLE
llvm12: inline asm: ambiguous instructions require an explicit suffix (could be 'pushw', or 'pushl')

### ISSUE THREAD

**Issue by mikdusan** [ziglang/zig#8123]
reproduction:
```
[llvm12⋅18-1] tiberius▸ zig test -target i386-linux-none test/stage1/behavior.zig

LLVM Emit Object... <inline asm>:1:3: error: ambiguous instructions require an explicit suffix (could be 'pushw', or 'pushl')
         push -28(%ebp)
         ^
LLVM ERROR: Error parsing inline asm
```

---
- change line 85 to `pushl` seems to fix. @LemonBoy?
https://github.com/ziglang/zig/blob/939f334221ad34618ceb59bc42ce9ef366a54bf9/lib/std/os/linux/i386.zig#L84-L100

- verified master on llvm11 produced test binary with i386.syscall6 symbol
```
[master⋅18-0] tiberius▸ zig test test/stage1/behavior.zig -target i386-linux-none
warning: created /Users/mike/project/zig/work/main/zig-cache/o/b736e77e94a22b4fb03b34f34b92071f/test but skipping execution because it is non-native

[master⋅18-0] tiberius▸ find zig-cache -name test
zig-cache/o/b736e77e94a22b4fb03b34f34b92071f/test

[master⋅18-0] tiberius▸ objdump -t zig-cache/o/b736e77e94a22b4fb03b34f34b92071f/test | grep sysc
004a80a0 l     F .text	00000019 std.os.linux.i386.syscall0
004bea00 l     F .text	00000021 std.os.linux.i386.syscall1
004a80c0 l     F .text	0000002e std.os.linux.i386.syscall2
004a8650 l     F .text	0000003a std.os.linux.i386.syscall3
004a7f60 l     F .text	00000045 std.os.linux.i386.syscall4
004ee6f0 l     F .text	0000006d std.os.linux.i386.syscall6
```
- looks like llvm12 became more strict depending on `push` form used and resulting ambiguity
- grep'd .zig codebase and the _ambiguous_ form doesn't appear anywhere else
```
[llvm12⋅18-1] tiberius▸ find * -name "*.zig" -print0 | xargs -0 grep '\\\\ *push'
lib/std/special/c.zig:                \\  push %%ebp
lib/std/special/c.zig:                \\  push %%ebx
lib/std/special/c.zig:                \\  push %%esi
lib/std/special/c.zig:                \\  push %%edi
lib/std/special/compiler_rt/aulldiv.zig:        \\  push   %%ebx
lib/std/special/compiler_rt/aulldiv.zig:        \\  push   %%esi
lib/std/special/compiler_rt/stack_probe.zig:                \\        push   %%rcx
lib/std/special/compiler_rt/stack_probe.zig:                \\        push   %%ecx
lib/std/special/compiler_rt/stack_probe.zig:                \\         push   %%rcx
lib/std/special/compiler_rt/stack_probe.zig:                \\         push   %%rax
lib/std/special/compiler_rt/stack_probe.zig:                \\         push   %%ecx
lib/std/special/compiler_rt/stack_probe.zig:                \\         push   %%eax
lib/std/special/compiler_rt/stack_probe.zig:                \\         push   %%rcx
lib/std/special/compiler_rt/stack_probe.zig:                \\         push   (%%rax)
lib/std/special/compiler_rt/stack_probe.zig:                \\         push   %%ecx
lib/std/special/compiler_rt/stack_probe.zig:                \\         push   (%%eax)
lib/std/special/compiler_rt/arm.zig:        \\ push {lr}
lib/std/special/compiler_rt/arm.zig:        \\ push {r4, lr}
lib/std/special/compiler_rt/arm.zig:        \\ push {lr}
lib/std/special/compiler_rt/arm.zig:        \\ push {r4, lr}
lib/std/special/compiler_rt/aullrem.zig:        \\  push   %%ebx
lib/std/os/linux/i386.zig:        \\ push %[arg6]
lib/std/os/linux/i386.zig:        \\ push %%ebp
```

**Comment by andrewrk** [comment]
[upstream bug](https://bugs.llvm.org/show_bug.cgi?id=49376)

**Comment by mikdusan** [comment]
- further reduction:
```
[llvm12⋅18-1] tiberius▸ zig build-obj reduction.zig -target i386-linux-none
<inline asm>:1:3: error: ambiguous instructions require an explicit suffix (could be 'pushw', or 'pushl')
         push -20(%ebp)
         ^
LLVM ERROR: Error parsing inline asm
```

- `reduction.zig`
```zig
export fn syscall6(
    number: u32,
    arg1: usize,
    arg2: usize,
    arg3: usize,
    arg4: usize,
    arg5: usize,
    arg6: usize,
) usize {
    return asm volatile (
        \\ push %[arg6]
        \\ push %%ebp
        \\ mov  4(%%esp), %%ebp
        \\ int  $0x80
        \\ pop  %%ebp
        \\ add  $4, %%esp
        : [ret] "={eax}" (-> usize)
        : [number] "{eax}" (number),
          [arg1] "{ebx}" (arg1),
          [arg2] "{ecx}" (arg2),
          [arg3] "{edx}" (arg3),
          [arg4] "{esi}" (arg4),
          [arg5] "{edi}" (arg5),
          [arg6] "rm" (arg6)
        : "memory"
    );
}

pub fn panic(msg: []const u8, error_return_trace: ?*@import("builtin").StackTrace) noreturn {
    while (true) {}
}
```

- command to emit `reduction.ll`
```sh
zig build-obj reduction.zig -target i386-linux-none -femit-llvm-ir -fno-emit-bin
```

- snip from `reduction.ll`
```llvm
; ModuleID = 'reduction'
source_filename = "reduction"
target datalayout = "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-f64:32:64-f80:32-n8:16:32-S128"
target triple = "i386-unknown-linux-unknown"

; Function Attrs: nobuiltin nounwind
define i32 @syscall6(i32 %0, i32 %1, i32 %2, i32 %3, i32 %4, i32 %5, i32 %6) #2 !dbg !45 {
Entry:
  %result = alloca i32, align 4
  %number = alloca i32, align 4
  %arg1 = alloca i32, align 4
  %arg2 = alloca i32, align 4
  %arg3 = alloca i32, align 4
  %arg4 = alloca i32, align 4
  %arg5 = alloca i32, align 4
  %arg6 = alloca i32, align 4
  store i32 %0, i32* %number, align 4
  call void @llvm.dbg.declare(metadata i32* %number, metadata !50, metadata !DIExpression()), !dbg !57
  store i32 %1, i32* %arg1, align 4
  call void @llvm.dbg.declare(metadata i32* %arg1, metadata !51, metadata !DIExpression()), !dbg !58
  store i32 %2, i32* %arg2, align 4
  call void @llvm.dbg.declare(metadata i32* %arg2, metadata !52, metadata !DIExpression()), !dbg !59
  store i32 %3, i32* %arg3, align 4
  call void @llvm.dbg.declare(metadata i32* %arg3, metadata !53, metadata !DIExpression()), !dbg !60
  store i32 %4, i32* %arg4, align 4
  call void @llvm.dbg.declare(metadata i32* %arg4, metadata !54, metadata !DIExpression()), !dbg !61
  store i32 %5, i32* %arg5, align 4
  call void @llvm.dbg.declare(metadata i32* %arg5, metadata !55, metadata !DIExpression()), !dbg !62
  store i32 %6, i32* %arg6, align 4
  call void @llvm.dbg.declare(metadata i32* %arg6, metadata !56, metadata !DIExpression()), !dbg !63
  %7 = load i32, i32* %number, align 4, !dbg !64
  %8 = load i32, i32* %arg1, align 4, !dbg !66
  %9 = load i32, i32* %arg2, align 4, !dbg !67
  %10 = load i32, i32* %arg3, align 4, !dbg !68
  %11 = load i32, i32* %arg4, align 4, !dbg !69
  %12 = load i32, i32* %arg5, align 4, !dbg !70
  %13 = load i32, i32* %arg6, align 4, !dbg !71
  %14 = call i32 asm sideeffect " push $7\0A push %ebp\0A mov  4(%esp), %ebp\0A int  $$0x80\0A pop  %ebp\0A add  $$4, %esp", "={eax},{eax},{ebx},{ecx},{edx},{esi},{edi},rm,~{memory},~{dirflag},~{fpsr},~{flags}"(i32 %7, i32 %8, i32 %9, i32 %10, i32 %11, i32 %12, i32 %13), !dbg !72
  store i32 %14, i32* %result, align 4, !dbg !72
  %15 = load i32, i32* %result, align 4, !dbg !73
  ret i32 %15, !dbg !73
}
```

**Comment by mikdusan** [comment]
looks related:
```
[llvm12⋅18-1] tiberius▸ zig test test/stage1/behavior.zig -lc -target i386-linux-musl
<inline asm>:11:2: error: ambiguous instructions require an explicit suffix (could be 'callw', or 'calll')
        call 1f
        ^
<inline asm>:14:2: error: ambiguous instructions require an explicit suffix (could be 'callw', or 'calll')
        call _start_c
        ^
```

**Comment by mikdusan** [comment]
and #8124 got me thinking...

- baseline
```sh
zig build-obj reduction.zig -target i386-linux-none
# error: ambiguous instructions...
```

- let's peek at llvm cpu features:
```sh
zig build-obj reduction.zig -target i386-linux-none --verbose-llvm-cpu-features
# error: ambiguous instructions...
```
```
name=reduction triple=i386-unknown-linux-unknown
name=reduction target_specific_cpu_args=pentium4
name=reduction target_specific_features=-16bit-mode,-32bit-mode,-3dnow,-3dnowa,-64bit,-adx,-aes,-amx-bf16,-amx-int8,-amx-tile,-avx,-avx2,-avx512bf16,-avx512bitalg,-avx512bw,-avx512cd,-avx512dq,-avx512er,-avx512f,-avx512ifma,-avx512pf,-avx512vbmi,-avx512vbmi2,-avx512vl,-avx512vnni,-avx512vp2intersect,-avx512vpopcntdq,-avxvnni,-bmi,-bmi2,-branchfusion,-cldemote,-clflushopt,-clwb,-clzero,+cmov,-cx16,+cx8,-enqcmd,-ermsb,-f16c,-false-deps-lzcnt-tzcnt,-false-deps-popcnt,-fast-11bytenop,-fast-15bytenop,-fast-7bytenop,-fast-bextr,-fast-gather,-fast-hops,-fast-lzcnt,-fast-scalar-fsqrt,-fast-scalar-shift-masks,-fast-shld-rotate,-fast-variable-shuffle,-fast-vector-fsqrt,-fast-vector-shift-masks,-fma,-fma4,-fsgsbase,-fsrm,+fxsr,-gfni,-hreset,-idivl-to-divb,-idivq-to-divl,-invpcid,-kl,-lea-sp,-lea-uses-ag,-lvi-cfi,-lvi-load-hardening,-lwp,-lzcnt,-macrofusion,+mmx,-movbe,-movdir64b,-movdiri,-mwaitx,+nopl,-pad-short-functions,-pclmul,-pconfig,-pku,-popcnt,-prefer-128-bit,-prefer-256-bit,-prefer-mask-registers,-prefetchwt1,-prfchw,-ptwrite,-rdpid,-rdrnd,-rdseed,-retpoline,-retpoline-external-thunk,-retpoline-indirect-branches,-retpoline-indirect-calls,-rtm,-sahf,-serialize,-seses,-sgx,-sha,-shstk,-slow-3ops-lea,-slow-incdec,-slow-lea,-slow-pmaddwd,-slow-pmulld,-slow-shld,-slow-two-mem-ops,+slow-unaligned-mem-16,-slow-unaligned-mem-32,-soft-float,+sse,+sse2,-sse3,-sse4.1,-sse4.2,-sse4a,-sse-unaligned-mem,-ssse3,-tbm,-tsxldtrk,-uintr,-use-aa,-use-glm-div-sqrt-costs,-vaes,-vpclmulqdq,+vzeroupper,-waitpkg,-wbnoinvd,-widekl,+x87,-xop,-xsave,-xsavec,-xsaveopt,-xsaves
```

- `-32bit_mode`? so what happens if we enable it? NO MORE LLVM ERROR!
```sh
zig build-obj reduction.zig -target i386-linux-none -mcpu=_i386+32bit_mode
# NO ERROR!
```

**Comment by andrewrk** [comment]
Thank you for the diagnosis @mikdusan! 

This issue is now solved with 68c00122b10a7364c52a59f56761942525498ce8 and 934b170d6769d1f632d3cf0ee314fb9e1249840a.
